File: 0838-design-linked-list/solution.java
/*
 * @lc app=leetcode id=707 lang=java
 *
 * [707] Design Linked List
 */

// @lc code=start
class MyLinkedList {
    private class Node {
        int val;
        Node next;

        Node() {
        }

        Node(int val) {
            this.val = val;
        }

        Node(int val, Node next) {
            this.val = val;
            this.next = next;
        }
    }

    private final Node sentinel;
    private int size;

    public MyLinkedList() {
        this.sentinel = new Node();
        this.size = 0;
    }

    public int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }

        Node p = this.sentinel;
        index++;
        while (index-- > 0) {
            p = p.next;
        }
        return p.val;
    }

    public void addAtHead(int val) {
        addAtIndex(0, val);
    }

    public void addAtTail(int val) {
        addAtIndex(size, val);
    }

    public void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }

        Node p = this.sentinel;

        while (index-- > 0) {
            p = p.next;
        }

        Node temp = p.next;
        p.next = new Node(val, temp);

        this.size++;
    }

    public void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }

        Node p = this.sentinel;

        while (index-- > 0) {
            p = p.next;
        }

        p.next = p.next.next;
        this.size--;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.get(index);
 * obj.addAtHead(val);
 * obj.addAtTail(val);
 * obj.addAtIndex(index,val);
 * obj.deleteAtIndex(index);
 */
// @lc code=end

File: 0838-design-linked-list/solution.py
class Node:
    
    def __init__(self, val : int = None, next : "Node | None" = None):
        self.val = val
        self.next = next

class MyLinkedList:

    def __init__(self):
        self.sentinel = Node()
        self.size = 0

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1

        cur = self.sentinel
        for _ in range(index):
           cur = cur.next

        return cur.next.val 
                

    def addAtHead(self, val: int) -> None:
        self.sentinel.next = Node(val, self.sentinel.next)
        self.size = self.size + 1

    def addAtTail(self, val: int) -> None:
        cur = self.sentinel
        while cur.next:
            cur = cur.next
        cur.next = Node(val)
        self.size = self.size + 1

    def addAtIndex(self, index: int, val: int) -> None:
        if index < 0 or index > self.size:
            return
        
        cur = self.sentinel
        for _ in range(index):
            cur = cur.next
        
        cur.next = Node(val, cur.next)
        self.size = self.size + 1

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return

        cur = self.sentinel
        for _ in range(index):
            cur = cur.next
        
        cur.next = cur.next.next
        self.size = self.size - 1


# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)
